google/cloud/compute/samples/disk_samples.cc

```diff
@@ -0,0 +1,138 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/compute/disks/v1/disks_client.h"
+#include "google/cloud/future.h"
+#include "google/cloud/internal/absl_str_cat_quiet.h"
+#include "google/cloud/internal/getenv.h"
+#include "google/cloud/internal/random.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/testing_util/example_driver.h"
+#include "absl/strings/str_split.h"
+#include <string>
+#include <vector>
+
+namespace {
+
+std::vector<std::pair<std::string, std::string>> SplitLabelsString(
+    std::string const& arg) {
+  std::vector<std::pair<std::string, std::string>> labels;
+  std::vector<std::string> pairs = absl::StrSplit(arg, ',');
+  std::transform(pairs.begin(), pairs.end(), std::back_inserter(labels),
+                 [](std::string const& p) {
+                   std::pair<std::string, std::string> l =
+                       absl::StrSplit(p, ':');
+                   return l;
+                 });
+  return labels;
+}
+
+void CreateEmptyDisk(std::vector<std::string> const& argv) {
+  if (argv.size() < 4) {
+    throw google::cloud::testing_util::Usage{
+        "compute-disk-create-empty-disk <project-id> <zone> <disk-name> "
+        "<disk-size-gb> "
+        "<labels>"};
+  }
+  //! [START compute_disk_create_empty_disk]
+  namespace compute_disks = ::google::cloud::compute_disks_v1;
+  namespace compute_proto = ::google::cloud::cpp::compute::v1;
+  using ::google::cloud::make_ready_future;
+  using ::google::cloud::StatusOr;
+
+  [](std::string const& project_id, std::string const& zone,
+     std::string const& disk_name, std::string const& disk_size_gb,
+     std::vector<std::pair<std::string, std::string>> const& labels) {
+    auto client =
+        compute_disks::DisksClient(compute_disks::MakeDisksConnectionRest());
+    compute_proto::Disk disk;
+    disk.set_name(disk_name);
+    disk.set_size_gb(disk_size_gb);
+    for (auto const& l : labels) {
+      (*disk.mutable_labels())[l.first] = l.second;
+    }
+    StatusOr<compute_proto::Disk> result =
+        client.InsertDisk(project_id, zone, disk)
+            .then([client, project_id, zone, disk_name](auto f) mutable {
+              StatusOr<compute_proto::Operation> create_result = f.get();
+              if (!create_result) {
+                return make_ready_future<StatusOr<compute_proto::Disk>>(
```

Comment:
We should not return futures inside a `.then()`. We should just return the `StatusOr<compute_proto::Disk>`

---

I think it works because this conversion from `future<future<T>>` -> `future<T>` is happening behind the scenes:
https://github.com/googleapis/google-cloud-cpp/blob/e3c2d5907ea572f3ff64c5f198b2aa48aeb6b109/google/cloud/internal/future_then_impl.h#L42-L44